### PR TITLE
BGDIINF_SB-1799: Implement post endpoint

### PR DESCRIPTION
**Pull request by IsabelleBzr**
Currently completed work on the post endpoint. I am now making a PR that I can make the AWS ECR into a separate branch and the PR will not be too big. The unittest will be extended at a later time in ticket [BGDIINF_SB-1816](https://jira.swisstopo.ch/browse/BGDIINF_SB-1816).

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -32,14 +32,14 @@ def add_cors_header(response):
...
-        logger.error('Origin=%s is not allowed', request.headers['Origin'])
-        abort(make_error_msg(403, 'Not allowed'))
+# @app.before_request
+# def validate_origin():
+#     if 'Origin' not in request.headers:
+#         logger.error('Origin header is not set')
+#         abort(make_error_msg(403, 'Not allowed'))
+#     if not re.match(ALLOWED_DOMAINS_PATTERN, request.headers['Origin']):
+#         logger.error('Origin=%s is not allowed', request.headers['Origin'])
+#         abort(make_error_msg(403, 'Not allowed'))
```
Comment (ltshb) [inline app/__init__.py:42]: Either remove this or add a `TODO CLEAN_UP`  comment to not forget about it.

File: app/helpers/response_generation.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+            'success': False, 'error': {
+                'code': code, 'message': msg
+            }
+        }),
+        code,
+        {
+            'Access-Control-Allow-Origin': '*',
+            'Access-Control-Allow-Methogs': 'POST',
+            'Access-Control-Allow-Headers': 'Content-Type'
+        }
```
Comment (ltshb) [inline app/helpers/response_generation.py:17]: CORS headers should be managed globally. See app/__init__.py:24

File: app/helpers/utils.py [hunk 3]
```diff
@@ -1,11 +1,17 @@
 import logging
 import logging.config
 import os
+import re
+from functools import wraps
 
+import defusedxml.ElementTree as ET
 import yaml
+from defusedxml.ElementTree import ParseError
```
Comment (ltshb) [inline app/helpers/utils.py:9]: Imports are not sorted, use `isort`

File: app/helpers/utils.py [hunk 4]
```diff
@@ -36,3 +44,47 @@ def get_logging_cfg():
 def init_logging():
     config = get_logging_cfg()
     logging.config.dictConfig(config)
+
+
+def prevent_erroneous_kml(kml_string):
+    kml_string = re.sub(r'on\w*=(".+?"|\'.+?\')', '', kml_string, flags=re.IGNORECASE)
```
Comment (ltshb) [inline app/helpers/utils.py:50]: A small docstring that explain what this function does, e.g. remove all attributes with `on` prefix and remove all `script` elements.

File: app/helpers/utils.py [hunk 5]
```diff
@@ -36,3 +44,47 @@ def get_logging_cfg():
...
+    if len(kml_string) > max_file_size:
+        error_msg = 'File size exceed %s bytes' % max_file_size
+        logger.error(error_msg)
+        abort(413, error_msg)
+    prevent_erroneous_kml(kml_string)
+    try:
+        ET.fromstring(kml_string)
+    except ParseError as err:
+        logger.error("Invalid kml file %s", err)
+        abort(415, 'Only valid KML file are accepted')
```
Comment (ltshb) [inline app/helpers/utils.py:88]: Here it should return a 400 Bad Request and not 415 Media Type Unsupported. Message should be 'Invalid KML file'

File: app/helpers/utils.py [hunk 6]
```diff
@@ -36,3 +44,47 @@ def get_logging_cfg():
...
+
+
+def validate_kml_string(kml_string):
+
+    max_file_size = 1024 * 1024 * 2
+
+    if len(kml_string) > max_file_size:
+        error_msg = 'File size exceed %s bytes' % max_file_size
+        logger.error(error_msg)
+        abort(413, error_msg)
```
Comment (ltshb) [inline app/helpers/utils.py:82]: Don't use old style  `%` string formatting (except for debug, info and warning logs). Also add the actual size to the log message. Might be good to print the size in MB for better readability.

File: app/helpers/utils.py [hunk 7]
```diff
@@ -36,3 +44,47 @@ def get_logging_cfg():
...
+
+        @wraps(func)
+        def wrapped(*args, **kwargs):
+            if request.content_type != content:
+                abort(make_error_msg(415, 'Unsupported Media Type'))
+            return func(*args, **kwargs)
+
+        return wrapped
+
+    return inner_decorator
```
Comment (ltshb) [inline app/helpers/utils.py:72]: Be more precise in the name of the method, it only validate the content type header not the content, so `validate_content_type` is better

**Review by ltshb — CHANGES_REQUESTED** [review]

File: app/helpers/utils.py [hunk 8]
```diff
@@ -36,3 +45,58 @@ def get_logging_cfg():
...
```
Comment (hansmannj) [inline app/helpers/utils.py:80]: Nice! :slightly_smiling_face:

**Review by hansmannj — APPROVED** [review]
:1st_place_medal: 
Already some useful unit tests :+1:

File: app/helpers/utils.py [hunk 9]
```diff
@@ -36,3 +45,58 @@ def get_logging_cfg():
...
+        kml_string,
+        flags=re.IGNORECASE | re.DOTALL
+    )
+    return kml_string
+
+
+def bytes_to_mb(bytes, to, b_size=1024):
+    a = {'k': 1, 'm': 2, 'g': 3, 't': 4, 'p': 5, 'e': 6}
+    r = float(bytes)
+    return bytes / (b_size**a[to])
```
Comment (ltshb) [inline app/helpers/utils.py:65]: Nice methods however I would change it as follow:

```python
def bytes_to(bytes, to, b_size=1024):
    a = {'kb': 1, 'mb': 2, 'gb': 3, 'tb': 4}
    r = float(bytes)
    return bytes / (b_size**a[to.lower()])
```
This made the usage clearer ;-) also peta und exa bytes are not needed.

File: app/helpers/utils.py [hunk 10]
```diff
@@ -36,3 +45,58 @@ def get_logging_cfg():
...
+    max_file_size = 1024 * 1024 * 2
+    file_size_in_mb = bytes_to_mb(max_file_size, 'm')
+    kml_string_mb = bytes_to_mb(len(kml_string), 'm')
+
+    if len(kml_string) > max_file_size:
+        error_msg = 'File size exceed {} MB. The actual file size is {} MB'.format(
+            file_size_in_mb, kml_string_mb
+        )
+        logger.error(error_msg)
+        abort(413, error_msg)
```
Comment (ltshb) [inline app/helpers/utils.py:94]: You don't need to generate the mb version, just do it on the fly in the error message formatting

`f"File size exceed {bytes_to(max_file_size, 'MB')} MB. The actual file size is {bytes_to(len(kml_string), 'MB')} MB"`
or 
`'File size exceed {} MB. The actual file size is {} MB'.format(
            bytes_to(max_file_size, 'MB'), bytes_to(len(kml_string), 'MB')
        )`

**Review by ltshb — APPROVED** [review]
Just minor improvements

**Pull request merged**